### ISSUE TITLE
Optimize Exporter.read method

### ISSUE THREAD

**Issue by bruth** [chop-dbhi/avocado#303]
I did a microbenchmark to evaluate the overhead of relatively standard formatters. A *standard* formatter is one that does not perform do any I/O such as a subquery or service call.

The `est. exc time` is the time it took for to get the first record and format it (this was included for curiosity and completeness since the original observation was an export timing out in the client). The `est. fmt time` is the amount of time is took to format all the rows once the iteration started.

### Setup
- using the CSV exporter
- 16 concepts with 61 fields

### Query

```sql
SELECT DISTINCT "production_patient"."id", "production_patient"."alias", "production_encounter"."age", "production_audiometrytestmethod"."name", "production_audiogramresult"."ear", "production_audiogramresult"."pta", "production_audiogramresult"."pta_worse", "production_audiogramresult"."f125", "production_audiogramresult"."f125_m", "production_audiogramresult"."f125_ml", "production_audiogramresult"."f125_n", "production_audiogramresult"."f125_v", "production_audiogramresult"."f250", "production_audiogramresult"."f250_m", "production_audiogramresult"."f250_ml", "production_audiogramresult"."f250_n", "production_audiogramresult"."f250_v", "production_audiogramresult"."f500", "production_audiogramresult"."f500_m", "production_audiogramresult"."f500_ml", "production_audiogramresult"."f500_n", "production_audiogramresult"."f500_v", "production_audiogramresult"."f750", "production_audiogramresult"."f750_m", "production_audiogramresult"."f750_ml", "production_audiogramresult"."f750_n", "production_audiogramresult"."f750_v", "production_audiogramresult"."f1000", "production_audiogramresult"."f1000_m", "production_audiogramresult"."f1000_ml", "production_audiogramresult"."f1000_n", "production_audiogramresult"."f1000_v", "production_audiogramresult"."f1500", "production_audiogramresult"."f1500_m", "production_audiogramresult"."f1500_ml", "production_audiogramresult"."f1500_n", "production_audiogramresult"."f1500_v", "production_audiogramresult"."f2000", "production_audiogramresult"."f2000_m", "production_audiogramresult"."f2000_ml", "production_audiogramresult"."f2000_n", "production_audiogramresult"."f2000_v", "production_audiogramresult"."f3000", "production_audiogramresult"."f3000_m", "production_audiogramresult"."f3000_ml", "production_audiogramresult"."f3000_n", "production_audiogramresult"."f3000_v", "production_audiogramresult"."f4000", "production_audiogramresult"."f4000_m", "production_audiogramresult"."f4000_ml", "production_audiogramresult"."f4000_n", "production_audiogramresult"."f4000_v", "production_audiogramresult"."f6000", "production_audiogramresult"."f6000_m", "production_audiogramresult"."f6000_ml", "production_audiogramresult"."f6000_n", "production_audiogramresult"."f6000_v", "production_audiogramresult"."f8000", "production_audiogramresult"."f8000_m", "production_audiogramresult"."f8000_ml", "production_audiogramresult"."f8000_n", "production_audiogramresult"."f8000_v"
FROM "production_patient" INNER JOIN "production_cohortpatient" ON ("production_patient"."id" = "production_cohortpatient"."patient_id") LEFT OUTER JOIN "production_encounter" ON ("production_patient"."id" = "production_encounter"."patient_id") LEFT OUTER JOIN "production_audiometrytest" ON ("production_encounter"."id" = "production_audiometrytest"."encounter_id") LEFT OUTER JOIN "production_audiometrytest_methods" ON ("production_audiometrytest"."id" = "production_audiometrytest_methods"."audiometrytest_id") LEFT OUTER JOIN "production_audiometrytestmethod" ON ("production_audiometrytest_methods"."audiometrytestmethod_id" = "production_audiometrytestmethod"."id") LEFT OUTER JOIN "production_audiogramtest" ON ("production_encounter"."id" = "production_audiogramtest"."encounter_id") LEFT OUTER JOIN "production_audiogramresult" ON ("production_audiogramtest"."id" = "production_audiogramresult"."test_id")
WHERE "production_cohortpatient"."cohort_id" IN (2)
ORDER BY "production_patient"."alias" ASC
```

### Results for `read`

This is the current implementation that forces distinct, checks for the limit and offset.

```
n=100
est. exc time: 9.56737613678
est. fmt time: 2.761

n=1000
est. exc time: 9.68558907509
est. fmt time: 4.265

n=10000
est. exc time: 9.97315597534
est. fmt time: 38.78

n=100000
DNF

n=623472
DNF
```

### Results for `fast_read`

This removes the force distinct, limit and offset logic from `read`.

```
n=100
est. fmt time: 0.834

n=1000
est. fmt time: 4.247

n=10000
est. fmt time: 36.84

n=100000
DNF

n=623472
DNF
```

### Results for `cache_read`

This removes the force distinct, limit and offset logic from `read` **and** caches the intermediate outputs from a formatter for the same set of inputs, e.g. `(formatter, inputs)` as the cache key.

This takes much more memory, however the time improvement is quite significant.

```
n=100
est. fmt time: 0.437

n=1000
est. fmt time: 0.278

n=10000
est. fmt time: 1.325

n=100000
est. fmt time: 9.45

n=623472
est. fmt time: 56.61
```

### Performance

- `read` - ~11919 µs/op
- `fast_read` - ~5427 µs/op
- `cache_read` - ~994 µs/op

### Observations

- Formatters are definitely a bottleneck during export.
- Although the `fast_read` time per op appears to be significant, this was influenced by the 100 row set which is not common for exports (in the results preview, yes).

### Caveats

- The `cache_read` method will only be valuable if there are repeated column data across rows.
- The amount of memory `cache_read` requires is not predicable.

**Comment by bruth** [comment]
Here is the code:

```python
v = DataView(...)
c = DataContext(...)
e = CSVExporter(v)
qs = v.apply(queryset=c.apply())

pows = [2, 3, 4, 5, 6]

def benchmark(p, reader):
    n = pow(10, p)

    it = qs[:n].query.get_compiler(qs.db).results_iter()

    t0 = time.time()

    for i, r in enumerate(reader(it)):
        if i == 0:
            t1 = time.time()

    t2 = time.time()

    print('n={}'.format(n))
    print('est. exc time: {}'.format(t1 - t0))
    print('est. fmt time: {}'.format(t2 - t1))
```

**Comment by bruth** [comment]
Did a first pass at #304 and removed `OrderedDict` and simplified the return values to be tuples only. Benchmarked against the `cache_read`.

```
n=100
est. fmt time: 0.618804216385

n=1000
est. fmt time: 0.0646691322327

n=10000
est. fmt time: 0.387552022934

n=100000
est. fmt time: 3.35769701004

n=623472
est. fmt time: 20.3275461197
```

**Comment by bruth** [comment]
The full result set took about ~850 MB of memory.

**Comment by bruth** [comment]
A threads + cache approach:

```
n=100
est. fmt time: 0.0152370929718

n=1000
est. fmt time: 0.0850520133972

n=10000
est. fmt time: 0.46654510498

n=100000
est. fmt time: 3.06137108803

n=623472
est. fmt time: 16.2542829514
```

**Comment by bruth** [comment]
The various read methods were implemented in #306, however various optimizations may still be required for larger data sets. These specifics will be opened as separate issues.
